### PR TITLE
Improve the documentation of `blocking`

### PR DESCRIPTION
**Pull request by farnz**
## Motivation

As seen in issue #618, experienced engineers get confused by the current documentation of `blocking`, because it blocks the whole task, not just a single future. While cleverer people than me are working on a more general solution, let's improve the documentation slightly to help out current users.

## Solution

Make the lack of a new thread explicit in the documentation, and point users of this function towards already existing mechanisms for creating new tasks if they need them. This is just a stop-gap while issue #629 is being worked on.

### CODE REVIEW

**Comment by tobz** [comment]
Can you reword this a little bit to explain why `blocking` isn't async, rather than talking about threads?

The situation that arises is pretty specific, and technically `blocking` does spawn a new thread/worker,  so we're really mincing words here.

**Comment by farnz** [comment]
How would you suggest rewording it?

My reason for talking about threads and tasks is to trigger the realisation that what `blocking` does is to claim the current task's executing thread for the duration of the blocking work; the spawning of a new thread is a side-effect if (and in the current implementation only if) there are no idle threads in the threadpool (and thus may never happen if the closure passed to `blocking` executes quickly enough, and there's an idle thread in the main pool waiting for an I/O event throughout).

My reason for putting it this way is that the implication that `blocking` spawns a new thread in the current documentation has misled at least 3 people into thinking that the closure runs on the new thread while leaving the rest of the task it's part of on the existing thread pool. I want to make it abundantly clear that this task is entirely blocked until the `blocking` section returns, and hence that the rest of the task won't run async until the closure completes, even if a new thread is spawned by the threadpool to handle async work.

**Comment by tobz** [comment]
So, I've been waffling back and forth on this, because the more I think about it, the more I think that what the comment says isn't so important as, probably, a better description in the docs about futures vs tasks. :P 

Could you just drop the sentence about the thread stuff?  That part is really the only thing that's not true and confuses more than it helps, I think.  As you said, we really just want people to know that chaining a blocking future directly will cause their task to block.

**Comment by farnz** [comment]
I've made another attempt at wording it, emphasising more that other futures in the same task will stop making progress while the closure is running.

The call out to the `select` combinator is because that's the one that'll most obviously do the "wrong" thing if you block for a long time. I'm not sure that my wording about "other futures" is correct, but I think it's good enough to push people to the realisation that it has to work this way because the scheduling atom in tokio is the task, not the individual future.

**Comment by tobz** [comment]
Great, yeah, this looks ideal now. 👍
